[Concept] Exercise - Switch and Enum (#694)
exercism/cpp: exercises/concept/troll-the-trolls/.docs/introduction.md

## Example usage

The specific term for this kind of enumeration is `scoped enumeration`.
The snippet below shows how to write a `DeckMaterial` `enumeration`.
Note the `enum class` keyword and the `;` at the end of the definition:

```cpp
enum class DeckMaterial {
    maple,
    bamboo,
    plastic
};
```
Now, look at a pricing function in the skate shop and take note of the scope resolution operator (`::`) specifying an `enumerator` from the `enumeration`:

```cpp
double deck_price(double base_price, DeckMaterial material) {
    if(material == DeckMaterial::plastic) {
        return base_price * 0.9;
    }
    return base_price * 1.3;
}
```

## Scoping

Imagine you have a second `enumeration` for wheel material:

```cpp
enum class WheelMaterial {
    steel,
    clay,
    plastic
};
```
Although the wheels and the deck can both be made of _plastic_, the two cannot be confused. 
They are different _types_:  `DeckMaterial` plastic and `WheelMaterial` plastic.
Each `enumeration` will have its `enumerators` in its own scope - its own `namespace`.
This is the reason they are called `scoped enumerations`.

~~~~exercism/advanced
## Unscoped Enumerations

> You might be thinking that with the name `scoped`, there would also be be `unscoped` enumerations -- and you would be correct.

Reviewer: ```suggestion
You might be thinking that with a name like `scoped`, there would also be  _`unscoped`_ enumerations -- and you would be correct.
```

`Unscoped enumerations` are becoming less popular because they all share the same global namespace.
Because of the sharing, you could not have two `unscoped enumerations` with the same `enumerators` like _plastic_ in the example above.

Also, `unscoped enumerations` implicitly convert to integers. 
Look at the example below for a surprising result:

```cpp
enum CitrusFruits {
    lemons,  // 0
    oranges, //1
};

enum IceCream {
    walnut,  // 0
    apples,  // 1
};

bool comparison{apples == oranges};
// => true

// Example from above:
bool comparison{DeckMaterial::plastic == WheelMaterial::plastic};
// => Does not compile!
```

If you want to convert `scoped enumerations` to integers you can use `static_cast<int>`.
~~~~

> [magic numbers]: https://en.wikipedia.org/wiki/Magic_number_(programming)

Reviewer: Since this is all one document, I think the reflink should move to the very bottom.

Author: This was generated with the `.md.tpl`. Would be nice if configlet would move links, but that might be too much?

## Switch Statements

Like other languages, C++ also provides a `switch` statement.
Switch statements are a shorter way to write long `if ... else if` statements.
To make a switch, we start by using the keyword `switch` followed by an integer.
We then declare each one of the conditions with the `case` keyword.

> We can also declare a `default` case, that will run when none of the previous `case` conditions matched.

Reviewer: ```suggestion
We can also declare a `default` case, that will run when none of the previous `case` conditions match.
```

Each case should end with a `break` (or a `return`) statement.

```cpp
int price{0};
int adults{3};
int kids{2};

switch (int group_size{adults + kids}) {
    case 1:
        price = 50;
        break;
    case 2:
        price = 70;
        break;
    default:
        price = group_size * 30;
}
```

## Fall-through

> One important thing about the switch construct is that the code will continue to execute until it is stopped by a `break` statement.

Reviewer: ```suggestion
One important thing about the switch construct is that the code will continue to execute until it is stopped by a `break` (or a `return`) statement.
```

This can lead to unexpected behavior.

```cpp
int adults{1};
int kids{0};
switch (int group_size{adults + kids}) {
    case 1:
        price = 50;
    case 2:
        price = 70;
    default:
        price = group_size * 30;
}
// price will be 30!
```

> The main use case for this feature is a statement, that has several labels.

Reviewer: ```suggestion
The main use case for this continued execution feature is a statement that has several labels.
```

> The same piece of code is executed for certain switch results.

Reviewer: ```suggestion
Multiple switch results can map to the same piece of code to be executed.
```

> This way - in a booking app the called function for the group sizes 2 and 3 can be the same.

Reviewer: ```suggestion
This way - in a booking app, for example - the called function for the group sizes 2 and 3 can be the same:
```
